_posts/2023-05-06-reverse-engineering-gba.markdown

```diff
@@ -0,0 +1,1062 @@
+
+To pass arguments into a function, the arguments are stored in registers before calling the function. `r0`-`r3` are used to pass arguments; if the function requires more than four arguments, any further arguments are pushed onto a place called the stack, which will be discussed later on.
+
+In the code below, a value is assigned to `r0` to pass as an argument into `FUN_080450f8`.
+```
+...
+adds r0,r4,#0x0
+bl FUN_080450f8
+```
+
+Once inside the function, the function can use the argument from `r0`.
+```
+FUN_080450f8
+
+ldr r0,[r0,#0x0]
+...
+```
+
+#### Return values
+If a function needs to return a value, the return value will be stored in `r0` just before the function returns. The caller function can use the return value in `r0` as needed.
+
+In function `FUN_08083654`, the function body assigns a value to `r0` before returning with `bx`. 
+```
+FUN_080450f8
+
+ldr r0,[r0,#0x0]
+bx lr
+```
+
+The caller can call the function, then retrieve the return value from `r0` to use for processing.
+```
+bl FUN_080450f8
+cmp r0,#0x1
+...
+```
+
+#### Call stack
+When a function is called, the caller is likely already using the registers to store values. There are only a handful of registers, and the function may also need those registers to do its work. Before the function can use the registers, it should save the existing values of registers it plans to use. When the function is finished, it should restore the saved values back to the registers so the caller doesn't lose its current state when it resumes execution.
+
+Registers `r0`-`r3` are designated as __scratch registers__, which are not saved by a function that uses them. Register `r4`-`r11` are __preserved registers__ whose values are saved and restored by the function. `lr` is also saved if the function calls other functions.
+
+Since a function can call another function, which can itself call another function and so on, every function must store and restore the register values at the proper times. This is accomplished using a place in memory called the call stack.
+
+The __call stack__, often shortened to __stack__, is a special place in memory used to save register values when functions are called. It is also used to store local variables if there are too many for the registers to hold. As the name implies, it is a last-in first-out (LIFO) data structure.
+
+In most games, the stack is located in memory at addresses 0x3000000-0x3007FFF. The top of the stack starts at address 0x3007FFF at the beginning of program execution. As values are pushed onto the stack, the top of the stack moves down. The address at the top of the stack is tracked using `r13`, often called the __stack pointer__ (`sp`).
+
+One of the main purposes of the function prologue is to save register values to the stack. Register values are pushed onto the top of the stack (i.e., the address of `sp`) using the `push` instruction. The instruction also decrements `sp` to reflect the new top of the stack.
```

Comment:
This is the first time a function prologue is mentioned and I'm not sure what it is.

Reply:
The parts of the function are introduced on line 373.